internal/firmware/revision.go

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package firmware
+
+import (
+	"fmt"
+	"strconv"
+)
+
+type Revision struct {
+	N int
+}
+
+func (r Revision) String() string {
+	if r.N == 0 {
+		return "unset"
+	}
+	if r.N < 0 {
+		return fmt.Sprintf("x%d", -r.N)
+	}
+	return strconv.Itoa(r.N)
+}
+
+func (r Revision) Unset() bool {
+	return r.N == 0
+}
+
+func (r Revision) Local() bool {
+	return r.N < 0
+}
+
+func (r Revision) Store() bool {
+	return r.N > 0
+}
+
+func (r Revision) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + r.String() + `"`), nil
+}
+
+func (r *Revision) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	return r.UnmarshalJSON([]byte(`"` + s + `"`))
+}
+
+func (r Revision) MarshalYAML() (interface{}, error) {
+	return r.String(), nil
+}
+
+func (r *Revision) UnmarshalJSON(data []byte) error {
+	if len(data) > 0 && data[0] == '"' && data[len(data)-1] == '"' {
+		parsed, err := ParseRevision(string(data[1 : len(data)-1]))
+		if err == nil {
+			*r = parsed
+			return nil
+		}
+	} else {
+		n, err := strconv.ParseInt(string(data), 10, 64)
+		if err == nil {
+			r.N = int(n)
+			return nil
+		}
+	}
+	return fmt.Errorf("invalid firmware revision: %q", data)
+}
+
+// ParseRevisions returns the representation in r as a revision.
+// See R for a function more suitable for hardcoded revisions.
+func ParseRevision(s string) (Revision, error) {
+	if s == "unset" {
+		return Revision{}, nil
+	}
+	if s != "" && s[0] == 'x' {
+		i, err := strconv.Atoi(s[1:])
+		if err == nil && i > 0 {
+			return Revision{-i}, nil
+		}
```

Comment:
so if s[0] is 'x' then we will never parse the revision number below. So we should just return a specific error message here.

Reply:
I hope this now makes more sense  - its a slightly confusing system if you have not seen this in action before :)